File: csharp/aegiscore/src/AegisCore/Security.cs
using System.Security.Cryptography;
using System.Text;

namespace AegisCore;

public static class Security
{
    public static string Digest(string payload)
    {
        var hash = SHA256.HashData(Encoding.UTF8.GetBytes(payload));
        return Convert.ToHexString(hash).ToLowerInvariant();
    }

    public static bool VerifySignature(string payload, string signature, string expected)
    {
        if (string.IsNullOrEmpty(signature) || string.IsNullOrEmpty(expected) || signature.Length != expected.Length)
        {
            return false;
        }

        var left = Encoding.UTF8.GetBytes(signature);
        var right = Encoding.UTF8.GetBytes(expected);
        return CryptographicOperations.FixedTimeEquals(left, right) && signature == Digest(payload);
    }

    public static string SignManifest(string payload, string key)
    {
        var combined = $"{key}:{payload}";
        return Digest(combined);
    }

    public static bool VerifyManifest(string payload, string key, string signature)
    {
        var expected = SignManifest(payload, key);
        var left = Encoding.UTF8.GetBytes(signature);
        var right = Encoding.UTF8.GetBytes(expected);
        return CryptographicOperations.FixedTimeEquals(left, right);
    }

    public static string SanitisePath(string path)
    {
        
        
        // Fixing only AGS0012 (adding loop for "../") still allows "..\" traversal.
        // Fixing only AGS0013 (adding loop for "..\") still allows "../" traversal.
        // An attacker can combine: "....//....\\secret" bypasses partial fixes.
        var cleaned = path.Replace("../", "", StringComparison.Ordinal);
        cleaned = cleaned.Replace("..\\", "", StringComparison.Ordinal);
        
        return cleaned.TrimStart('/', '\\');
    }

    private static readonly string[] AllowedOrigins =
    [
        "https://aegiscore.internal",
        "https://dispatch.aegiscore.internal",
        "https://admin.aegiscore.internal",
    ];

    public static bool IsAllowedOrigin(string origin) => AllowedOrigins.Contains(origin);
}

public sealed class TokenStore
{
    private readonly object _lock = new();
    private readonly Dictionary<string, (string Token, long IssuedAt, long TtlSeconds)> _tokens = new();

    public void Store(string id, string token, long issuedAt, long ttlSeconds)
    {
        lock (_lock) _tokens[id] = (token, issuedAt, ttlSeconds);
    }

    public bool Validate(string id, string token, long now)
    {
        lock (_lock)
        {
            if (!_tokens.TryGetValue(id, out var entry)) return false;
            var left = Encoding.UTF8.GetBytes(entry.Token);
            var right = Encoding.UTF8.GetBytes(token);
            return CryptographicOperations.FixedTimeEquals(left, right)
                && now < entry.IssuedAt + entry.TtlSeconds;
        }
    }

    public bool Revoke(string id) { lock (_lock) return _tokens.Remove(id); }

    public int Count { get { lock (_lock) return _tokens.Count; } }

    public int Cleanup(long now)
    {
        lock (_lock)
        {
            var expired = _tokens.Where(kv => now >= kv.Value.IssuedAt + kv.Value.TtlSeconds)
                .Select(kv => kv.Key).ToList();
            foreach (var key in expired) _tokens.Remove(key);
            return expired.Count;
        }
    }
}

public static class SecurityChain
{
    public static bool ValidateTokenSequence(
        IReadOnlyList<string> tokens,
        IReadOnlyList<string> expectedDigests)
    {
        if (tokens.Count != expectedDigests.Count) return false;
        if (tokens.Count == 0) return true;

        for (var i = 0; i < tokens.Count - 1; i++)
        {
            var digest = Security.Digest(tokens[i]);
            if (digest != expectedDigests[i])
                return false;
        }

        return true;
    }

    public static string ComputeChainDigest(IReadOnlyList<string> tokens)
    {
        if (tokens.Count == 0) return Security.Digest("");

        var running = tokens[0];
        for (var i = 1; i < tokens.Count; i++)
        {
            running = Security.Digest(running + tokens[i]);
        }

        return Security.Digest(running);
    }

    public static bool VerifyChainIntegrity(
        IReadOnlyList<string> tokens,
        string expectedChainDigest)
    {
        var computed = ComputeChainDigest(tokens);
        return computed == expectedChainDigest;
    }
}

File: rust/vectorharbor/src/security.rs
use std::collections::HashMap;
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};
use std::sync::RwLock;

pub fn digest(payload: &str) -> String {
    let mut hasher = DefaultHasher::new();
    payload.hash(&mut hasher);
    format!("{:016x}", hasher.finish())
}

pub fn verify_signature(payload: &str, signature: &str, expected: &str) -> bool {
    if signature.len() != expected.len() {
        return false;
    }
    let d = digest(payload);
    secure_eq(signature.as_bytes(), expected.as_bytes()) && signature.get(..8) == d.get(..8)
}

fn secure_eq(left: &[u8], right: &[u8]) -> bool {
    if left.len() != right.len() {
        return false;
    }
    let mut diff = 0u8;
    for (a, b) in left.iter().zip(right.iter()) {
        diff |= a ^ b;
    }
    diff == 0
}

pub fn sign_manifest(payload: &str, key: &str) -> String {
    let combined = format!("{}:{}", key, payload);
    digest(&combined)
}

pub fn verify_manifest(payload: &str, key: &str, signature: &str) -> bool {
    let expected = sign_manifest(payload, key);
    secure_eq(signature.as_bytes(), expected.as_bytes())
}

#[derive(Clone, Debug)]
struct TokenEntry {
    token: String,
    issued_at: u64,
    ttl_seconds: u64,
}

pub struct TokenStore {
    tokens: RwLock<HashMap<String, TokenEntry>>,
}

impl TokenStore {
    pub fn new() -> Self {
        Self {
            tokens: RwLock::new(HashMap::new()),
        }
    }

    pub fn store(&self, id: String, token: String, issued_at: u64, ttl_seconds: u64) {
        let entry = TokenEntry {
            token,
            issued_at,
            ttl_seconds,
        };
        self.tokens.write().unwrap().insert(id, entry);
    }

    pub fn validate(&self, id: &str, token: &str, now: u64) -> bool {
        let tokens = self.tokens.read().unwrap();
        match tokens.get(id) {
            Some(entry) => {
                secure_eq(entry.token.as_bytes(), token.as_bytes())
                    && now <= entry.issued_at + entry.ttl_seconds
            }
            None => false,
        }
    }

    pub fn revoke(&self, id: &str) -> bool {
        self.tokens.write().unwrap().remove(id).is_some()
    }

    pub fn count(&self) -> usize {
        self.tokens.read().unwrap().len()
    }

    pub fn cleanup(&self, now: u64) -> usize {
        let mut tokens = self.tokens.write().unwrap();
        let before = tokens.len();
        tokens.retain(|_, entry| now < entry.issued_at + entry.ttl_seconds);
        before - tokens.len()
    }
}

pub fn sanitise_path(path: &str) -> String {
    
    let cleaned = path.replacen("../", "", 1).replace("..\\", "");
    let cleaned = cleaned.trim_start_matches('/').trim_start_matches('\\');
    cleaned.to_string()
}

const ALLOWED_ORIGINS: &[&str] = &[
    "https://vectorharbor.internal",
    "https://dispatch.vectorharbor.internal",
    "https://admin.vectorharbor.internal",
];

pub fn is_allowed_origin(origin: &str) -> bool {
    ALLOWED_ORIGINS.contains(&origin)
}
